File: samples/client/3_0_3_unit_test/java/src/test/java/org/openapijsonschematools/client/components/schemas/EnumsInPropertiesTest.java
package org.openapijsonschematools.client.components.schemas;

import org.junit.Assert;
import org.junit.Test;
import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags;
import org.openapijsonschematools.client.configurations.SchemaConfiguration;
import org.openapijsonschematools.client.exceptions.ValidationException;
import org.openapijsonschematools.client.schemas.MapMaker;
import org.openapijsonschematools.client.schemas.validation.JsonSchema;
import org.openapijsonschematools.client.schemas.validation.PathToSchemasMap;
import org.openapijsonschematools.client.schemas.validation.ValidationMetadata;

import java.util.Arrays;
import java.util.List;
import java.util.AbstractMap;
import java.util.LinkedHashSet;

public class EnumsInPropertiesTest {
    static final SchemaConfiguration configuration = new SchemaConfiguration(JsonSchemaKeywordFlags.ofNone());
    static final ValidationMetadata validationMetadata = new ValidationMetadata(
            List.of("args[0"),
            configuration,
            new PathToSchemasMap(),
            new LinkedHashSet<>()
    );

    @Test
    public void testWrongBarValueFails() {
        // wrong bar value
        final var schema = EnumsInProperties.EnumsInProperties1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<>(
                    "foo",
                    "foo"
                ),
                new AbstractMap.SimpleEntry<>(
                    "bar",
                    "bart"
                )
            ),
            validationMetadata
        ));
    }

    @Test
    public void testWrongFooValueFails() {
        // wrong foo value
        final var schema = EnumsInProperties.EnumsInProperties1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<>(
                    "foo",
                    "foot"
                ),
                new AbstractMap.SimpleEntry<>(
                    "bar",
                    "bar"
                )
            ),
            validationMetadata
        ));
    }

    @Test
    public void testMissingAllPropertiesIsInvalidFails() {
        // missing all properties is invalid
        final var schema = EnumsInProperties.EnumsInProperties1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
            ),
            validationMetadata
        ));
    }

    @Test
    public void testBothPropertiesAreValidPasses() {
        // both properties are valid
        final var schema = EnumsInProperties.EnumsInProperties1.getInstance();
        schema.validate(
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<>(
                    "foo",
                    "foo"
                ),
                new AbstractMap.SimpleEntry<>(
                    "bar",
                    "bar"
                )
            ),
            configuration
        );
    }

    @Test
    public void testMissingOptionalPropertyIsValidPasses() {
        // missing optional property is valid
        final var schema = EnumsInProperties.EnumsInProperties1.getInstance();
        schema.validate(
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<>(
                    "bar",
                    "bar"
                )
            ),
            configuration
        );
    }

    @Test
    public void testMissingRequiredPropertyIsInvalidFails() {
        // missing required property is invalid
        final var schema = EnumsInProperties.EnumsInProperties1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<>(
                    "foo",
                    "foo"
                )
            ),
            validationMetadata
        ));
    }
}

File: samples/client/3_0_3_unit_test/kotlin/src/test/kotlin/org/openapijsonschematools/client/components/schemas/EnumsInPropertiesTest.kt
package org.openapijsonschematools.client.components.schemas

import org.openapijsonschematools.client.configurations.JsonSchemaKeyword
import org.openapijsonschematools.client.configurations.SchemaConfiguration
import org.openapijsonschematools.client.exceptions.ValidationException

import kotlin.test.Test
import kotlin.test.assertFailsWith

class EnumsInPropertiesTest {
    companion object {
        val configuration = SchemaConfiguration(disabledKeywordFlags=setOf(JsonSchemaKeyword.FORMAT))
    }

    @Test
    fun testWrongBarValueFails() {
        // wrong bar value
        val schema = EnumsInProperties.EnumsInProperties1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "foo",
                            "foo"
                        ),
                        Pair(
                            "bar",
                            "bart"
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testWrongFooValueFails() {
        // wrong foo value
        val schema = EnumsInProperties.EnumsInProperties1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "foo",
                            "foot"
                        ),
                        Pair(
                            "bar",
                            "bar"
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testMissingAllPropertiesIsInvalidFails() {
        // missing all properties is invalid
        val schema = EnumsInProperties.EnumsInProperties1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf<Any?, Any?>(
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testBothPropertiesAreValidPasses() {
        // both properties are valid
        val schema = EnumsInProperties.EnumsInProperties1.getInstance()
        schema.validate(
            mapOf(
                Pair(
                    "foo",
                    "foo"
                ),
                Pair(
                    "bar",
                    "bar"
                )
            ),
            configuration
        )
    }

    @Test
    fun testMissingOptionalPropertyIsValidPasses() {
        // missing optional property is valid
        val schema = EnumsInProperties.EnumsInProperties1.getInstance()
        schema.validate(
            mapOf(
                Pair(
                    "bar",
                    "bar"
                )
            ),
            configuration
        )
    }

    @Test
    fun testMissingRequiredPropertyIsInvalidFails() {
        // missing required property is invalid
        val schema = EnumsInProperties.EnumsInProperties1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "foo",
                            "foo"
                        )
                    ),
                    configuration
                )
            }
        )
    }
}
